dash/include/dash/memory/MemorySpace.h

```diff
@@ -1,18 +1,268 @@
+struct memory_space_local_domain_tag {
+};
+
+struct memory_space_host_tag {
+};
+struct memory_space_hbw_tag {
+};
+struct memory_space_pmem_tag {
+};
+
+template <class MemSpace>
+struct memory_space_traits {
+  using memory_space_domain_category =
+      typename MemSpace::memory_space_domain_category;
+  using memory_space_type_category =
+      typename MemSpace::memory_space_type_category;
+
+  /**
+   * Whether the memory space type is specified for global address space.
+   */
+  using is_global = std::integral_constant<
+      bool, std::is_same<memory_space_domain_category,
+                         memory_space_global_domain_tag>::value>;
+  /**
+   * Whether the memory space type is specified for local address space.
+   * As arbitrary address space domains can be defined, this trait is not
+   * equivalent to `!is_global`.
+   */
+  using is_local = std::integral_constant<
+      bool, std::is_same<memory_space_domain_category,
+                         memory_space_local_domain_tag>::value>;
+  using void_pointer =
+      typename std::conditional<is_global::value, dash::GlobPtr<void, MemSpace>,
+                                void*>::type;
+};
+
+/**
+ * The \c MemorySpace concept follows the STL \c std::pmr::memory_resource.
+ *
+ * Unlinke shared memory systems where process memory and virtual address
+ * translation is realized as system functions and hardware components,
+ * PGAS memory resources are also responsible for maintaining and propagating
+ * the size and structure of their underlying local memory ranges.
+ *
+ * Consequently, the memory resource concept is extended by methods and
+ * type definitions that are required to maintain global pointer arithmetics
+ * and named Memory Space in DASH.
+ *
+ * STL Reference of \c std::pmr::memory_resource:
+ * http://en.cppreference.com/w/cpp/memory/memory_resource
+ *
+ * \note
+ *
+ * Defining \c MemorySpace as class template seems to contradict the
+ * intention to use it as polymorphic base class as subclass types of
+ * \c MemorySpace<T> and \c MemorySpace<U> have incompatible polymorphic
+ * base classes.
+ * This looks like the kind of unintentional type incompatibility that lead
+ * to the introduction of polymorphic allocators.
+ * However, the template parameter specifies the memory space domain and
+ * two types of \c MemorySpace are in fact incompatible if they do not refer
+ * to the same domain, like global and local address space.
+ *
+ * \todo
+ *
+ * Clarify:
+ * The STL memory_resource concept assumes that any type can be
+ * allocated with its size specified in number of bytes so untyped
+ * allocation with void pointers is practicable.
+ * However, the DART allocation routines are typed for correctness
+ * considerations and to optimize communication.
+ * If the memory space concept complies to the STL memory_resource
+ * interface, only DART_TYPE_BYTE instead of the actual value type
+ * is available. This would therefore harm stability and performance.
```

Comment:
Note: the type argument in DART allocation functions is not used and you can safely pass `DART_TYPE_BYTE` to them. It do not affect alignment, which will be taken care of later (DART memory allocation overhaul). In fact, passing the type to the underlying MPI implementation only affects the indexing in access functions (which is in bytes for us, for reasons).